Fix up element/elements/section/sections to all contain correct tests

luke-hill opened the pull request:
As well as that I've introduced some additional xpath tests where appropriate to duplicate the test coverage where we perhaps were a bit light.

luke-hill commented:
Ping @TheMetalCode

luke-hill commented:
As I think you're back from hols I may as well ping you as well @tgaff

tgaff commented:
Back in just a couple days actually.  I'll try to take a look soon.


On Thu, Apr 19, 2018, 17:04 Luke Hill <notifications@github.com> wrote:

> As I think you're back from hols I may as well ping you as well @tgaff
> <https://github.com/tgaff>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/natritmeyer/site_prism/pull/236#issuecomment-382647220>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB088H85m9vjzEzkistd5UnKVFsqHvw6ks5tqET_gaJpZM4TWCO4>
> .
>

luke-hill commented on spec/element_spec.rb:
```diff
@@ -3,105 +3,68 @@
-    expect(page).to respond_to :has_no_thing?
-  end
+    describe '#expected_elements' do
+      subject { page.all_there? }
```
Yes it is whoops :)

luke-hill commented on spec/element_spec.rb:
```diff
@@ -3,105 +3,68 @@
-  it 'should know if all mapped elements are on the page' do
-    class PageWithAFewElements < SitePrism::Page
+  context 'with a css element' do
+    class Page < SitePrism::Page
```
Yes we could probably. I'll refactor this. Good spot. (I'll do this for both)

luke-hill commented:
I've fixed up this branch and fixed up most of your comments @tgaff. I've also added extra tests where appropriate from the work from @ricmatsui and kept as much intact as possible.

In adding the extra tests. RSpec started to mix the two shared contexts together (`expected_elements` became duplicated). So what I did is called the classes distinct names, and then referenced the verbose classname in the classmethod description by doing `klass.expected_elements`

luke-hill commented:
I'd like to merge this PR in next (As this is the one that is most likely to conflict with other stuff), as it's the last "big" refactor of spec code (There is another PR in the works which refactors a load of cucumber code as well). And then cut 2.13 as the last big "refactor heavy" release

cc/ @tgaff
